src/material-experimental/dialog/dialog-container.ts

```diff
@@ -0,0 +1,217 @@
+}
+
+
+/**
+ * Internal component that wraps user-provided dialog content.
+ * @docs-private
+ */
+@Component({
+  moduleId: module.id,
+  selector: 'cdk-dialog-container',
+  template: `<ng-template cdkPortalOutlet></ng-template>`,
+  styles: [`
+    cdk-dialog-container {
+      background: white;
+      border-radius: 5px;
+      display: block;
+      padding: 10px;
+    }
+  `],
+  encapsulation: ViewEncapsulation.None,
+  preserveWhitespaces: false,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  animations: [
+    trigger('dialog', [
+      state('enter', style({ opacity: 1 })),
+      state('exit, void', style({ opacity: 0 })),
+      transition('* => *', animate(225)),
+    ])
+  ],
+  host: {
+    '[@dialog]': '_state',
+    '(@dialog.start)': '_onAnimationStart($event)',
+    '(@dialog.done)': '_onAnimationDone($event)',
+  },
+})
+export class CdkDialogContainer extends BasePortalHost {
+  /** State of the dialog animation. */
+  _state: 'void' | 'enter' | 'exit' = 'enter';
+
+  /** Element that was focused before the dialog was opened. Save this to restore upon close. */
+  private _elementFocusedBeforeDialogWasOpened: HTMLElement | null = null;
+
+   /** The class that traps and manages focus within the dialog. */
+  private _focusTrap = this._focusTrapFactory.create(this._elementRef.nativeElement, false);
+
+  // tslint:disable-next-line:no-host-decorator-in-concrete
+  @HostBinding('attr.aria-label') private get _ariaLabel() {
+    return this._config.ariaLabel || null;
+  }
+
+  // tslint:disable-next-line:no-host-decorator-in-concrete
+  @HostBinding('attr.aria-describedby')
+  private get ariaDescribedBy() { return this._config ? this._config.ariaDescribedBy : null; }
+
+  // tslint:disable-next-line:no-host-decorator-in-concrete
+  @HostBinding('attr.role') private get role() { return this._config ? this._config.role : null; }
+
+  // tslint:disable-next-line:no-host-decorator-in-concrete
+  @HostBinding('attr.tabindex') private get tabindex() { return -1; }
+
+  /** The portal host inside of this container into which the dialog content will be loaded. */
+  @ViewChild(PortalHostDirective) private _portalHost: PortalHostDirective;
+
+  /** A subject emitting before the dialog enters the view. */
+  _beforeEnter: Subject<void> = new Subject();
+
+  /** A subject emitting after the dialog enters the view. */
+  _afterEnter: Subject<void> = new Subject();
+
+  /** A subject emitting before the dialog exits the view. */
+  _beforeExit: Subject<void> = new Subject();
+
+  /** A subject emitting after the dialog exits the view. */
+  _afterExit: Subject<void> = new Subject();
+
+  /** The dialog configuration. */
+  _config: DialogConfig;
+
+  constructor(
+    private _elementRef: ElementRef,
+    private _focusTrapFactory: FocusTrapFactory,
+    private _changeDetectorRef: ChangeDetectorRef,
+    @Optional() @Inject(DOCUMENT) private _document: any) {
+    super();
+
+    this._afterEnter.subscribe(() => {
```

Comment:
Rather than having it subscribe to its own observables, it might be better to just do the calls at the appropriate time.

Reply:
Personally I find it a bit more readable to see that these actions happen in response to "lifecycle hooks".  But I could see arguements for both.

@jelbourn thoughts?

Reply:
I'd agree w/ @crisbeto; it feels more natural to do this work in the callbacks you've set up. 

Also the inside of each of these should really be their own methods like `_applyAutoFocus` or something like that